src/cluster.c

```diff
@@ -5922,44 +6115,129 @@ NULL
+             * before SETSLOT has been replicated to B'. If B crashes here,
+             * B' will be in importing state and the slot will have no owner.
+             * To help mitigate this issue, we enforce the following order
+             * for slot migration finalization such that the replicas will
+             * finalize the slot ownership before this primary:
+             *
+             * 1. Client C issues SETSLOT n NODE B against node B
+             * 2. Node B replicates SETSLOT n NODE B to all of its
+             *    replicas, such as B', B'', etc
+             * 3. On replication completion, node B executes SETSLOT
+             *    n NODE B and returns control back to client C
+             * 4. The following steps can happen in parallel
+             * a. Client C issues SETSLOT n NODE B against node A
+             * b. node B gossips its new slot ownership to the cluster
+             *    including A, A', etc
+             *
+             * Where A is the source primary and B is the destination primary.
+             *
+             * This enforcement is implemented via a new phase called `REPLICAONLY'`
+             * in the setslot subcommand. */
+
+            n = clusterLookupNode(c->argv[4]->ptr, sdslen(c->argv[4]->ptr));
+
+            if (!n) {
+                addReplyErrorFormat(c,"Unknown node %s", (char*)c->argv[4]->ptr);
+                return;
+            }
+            if (nodeIsSlave(n)) {
+                addReplyError(c,"Target node is not a master");
+                return;
+            }
+            /* If this hash slot was served by 'myself' before to switch
+             * make sure there are no longer local keys for this hash slot. */
+            if (server.cluster->slots[slot] == myself && n != myself) {
+                if (countKeysInSlot(slot) != 0) {
+                    addReplyErrorFormat(c,
+                        "Can't assign hashslot %d to a different node "
+                        "while I still hold keys for this hash slot.", slot);
+                    return;
+                }
+            }
+
+            /* Replicate the command to the replicas if there are any */
+            if ((server.cluster->importing_slots_from[slot] != NULL) &&
+                (nodeIsMaster(myself) != 0) &&
+                (myself->numslaves != 0))
+            {
+                /* Remove the last "REPLICAONLY" token so the command
+                 * can be applied as the real "SETSLOT" command on the
+                 * replicas. */
+                serverAssert(c->argc == 6);
+                c->argc--;
+                decrRefCount(c->argv[5]);
+                c->argv[5] = NULL;
+
+                /* We are a primary and this is the first time we see this "setslot"
+                 * command. Force-replicate the setslot command to all of our replicas
+                 * first and only on success will we handle the command.
```

Comment:
Is this right? I though SETSLOT REPLICAONLY causes repliction only and has no effect on the node itself. Or what do you mean by "only on success will we handle the command"?

Another question: In the command SETSLOT ... NODE (without REPLICAONLY), is the command replicated or not? If not (and if an old redis-cli is used which doesn't send the new REPLICAONLY command) do we rely on the cluster bus propagating the information to replicas instead?

Reply:
This section is now re-written. Can you please review again?